packages/checkout/registry/index.js

```diff
@@ -0,0 +1,56 @@
+let checkoutFilters = {};
+
+/**
+ * Register filters for a specific extension.
+ *
+ * @param {string} namespace Name of the extension namespace.
+ * @param {Object} filters   Object of filters for that namespace. Each key of
+ *                           the object is the name of a filter.
+ */
+export const __experimentalRegisterCheckoutFilters = ( namespace, filters ) => {
+	checkoutFilters = {
+		...checkoutFilters,
+		[ namespace ]: filters,
+	};
+};
+
+/**
+ * Get all filters with a specific name.
+ *
+ * @param {string} filterName   Name of the filter to search for.
+ * @return {Function[]} Array of functions that are registered for that filter
+ *                      name.
+ */
+const getCheckoutFilters = ( filterName ) => {
+	const namespaces = Object.keys( checkoutFilters );
+	const filters = namespaces
+		.map( ( namespace ) => checkoutFilters[ namespace ][ filterName ] )
+		.filter( Boolean );
+	return filters;
+};
+
+/**
+ * Apply a filter.
+ *
+ * @param {string}   filterName   Name of the filter to apply.
+ * @param {any}      defaultValue Default value to filter.
+ * @param {Object}   [args]       Arguments to pass to registered functions.
+ * @param {Function} [validate]   Function that needs to return true when the
+ *                                filtered value is passed in order for the
+ *                                filter to be applied.
+ * @return {any} Filtered value.
+ */
+export const __experimentalApplyCheckoutFilter = (
+	filterName,
+	defaultValue,
+	args = {},
+	validate = () => true
+) => {
+	const filters = getCheckoutFilters( filterName );
+	let value = defaultValue;
+	filters.forEach( ( filter ) => {
+		const newValue = filter( value, args );
```

Comment:
This is too dangerous to just execute like this, any malfunction in the 3PD code would break our block.
You can test this by changing `extensions` to `extension` in WCS code, the whole cart block is now broken.

The error can be at runtime or can happen with a version mismatch (so the 3PD import something that is no longer available) or just anything.

A very important principle when developing the new extensibility points is to not trust the developer.

The filter call should be wrapped in a try/catch closure that will skip that particular filter something wrong happens.

Skipping filters is only possible on raw values, we can't know for sure if an element is faulty or not until we evaluate it, meaning we would break other extensions.

We could wrap our `__experimentalApplyCheckoutFilter` call in a BlockErrorBoudary but that would still neglect the rest of filters.

We can also nest all filters into each other somehow, but I'm not sure how would that happen.

I'd suggest we skip element for now and only have raw values until we have a strong case for elements and then we can think about it very well.

Reply:
Good points, thanks for the feedback. In 660c0266af2494d2eb48469c89175b16b1b349ae I wrapped the filter function in a `try { ... } catch { ... }` block and in 094eea75ef254eeb9f61354b3bcbdfb9041c839a I updated the validation so only strings are accepted. Let me know if changes make sense.